### PR TITLE
Updated Badge Component

### PR DESCRIPTION
**Pull request by cameron-eyds**
*Issue #:* /bcgov/entity#19503

*Description of changes:*
* Created small badge component which takes two data sets, compares them and renders when there is differences. 
* Deep comparison function will normalize strings, take numbers, booleans, objects and arrays, including nested objects and arrays (see unit tests).
* Lots of unit tests
* Included screens for example implementation:

(baseline would take the snapshot value, just hardcoded this for the example)
Screenshot 2024-02-16 at 9 29 46 AM
Screenshot 2024-02-16 at 9 29 26 AM
Screenshot 2024-02-16 at 9 29 20 AM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://bcregistry-assets-dev--pr-1739-04mdadnb.web.app

File: ppr-ui/src/components/common/UpdatedBadge.vue [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+const props = withDefaults(defineProps<{
+  action?: string,
+  baseline: BaseDataUnionIF,
+  currentState: BaseDataUnionIF
+}>(), {
+  action: 'CORRECTED',
+  baseline: null,
+  currentState: null
+})
```
Comment (cameron-eyds) [inline ppr-ui/src/components/common/UpdatedBadge.vue:26]: @dimak1 Small heads up here, when deconstructing props to apply defaults as in MhrTransportPermit line 258, it has reactivity implications. I followed that same approach initially and was losing reactivity on those props. 

Screenshot 2024-02-16 at 11 18 39 AM

Comment (cameron-eyds) [reply ppr-ui/src/components/common/UpdatedBadge.vue:26]: Might just be when trying to type our props, as it states in the documentation there. 
Either way, not sure if it was an issue in the TransportPermit, just something to be aware of!

**Review by doug-lovett — APPROVED** [review]
Very nice.

File: ppr-ui/src/utils/utilities.ts [hunk 2]
```diff
@@ -81,3 +83,31 @@ export const filterDuplicates = (list: Array<any>, filterBy: string) => {
...
+ * @param {*} current - The second value to compare.
+ * @returns {boolean} - Returns true if the values are different, false if they are equal.
+ */
+export const deepChangesComparison = (base: BaseDataUnionIF, current: BaseDataUnionIF): boolean => {
+  // Object safety-check
+  const isObject = (value) => typeof value === 'object' && value !== null
+  // String normalization safety-check: Case Insensitive
+  const caseInsensitiveStringCompare = (val1, val2) => {
+    if (typeof val1 === 'string' && typeof val2 === 'string') {
+      return val1.toUpperCase() !== val2.toUpperCase()
```
Comment (dimak1) [inline ppr-ui/src/utils/utilities.ts:100]: So we are ignoring the case? Liz was ok with that?

**Review by dimak1 — APPROVED** [review]
The only thing I can think of is performance of the deep comparison. In case we have many badges (Review page) and it takes long to load, we can future proof by making it async. It is not a requirement however, so time will show.
Nice work!

File: ppr-ui/src/utils/utilities.ts (hunk 2 above)
Comment (cameron-eyds) [reply ppr-ui/src/utils/utilities.ts:100]: It was an explicit requirement in corrections, so it should be the case here too yea. 
If they have issue, we can option this.

**Comment by cameron-eyds** [comment]
> The only thing I can think of is performance of the deep comparison. In case we have many badges (Review page) and it takes long to load, we can future proof by making it async. It is not a requirement however, so time will show. Nice work!

Performance impact is negligible but it depends on the implementation. It hinges on the length and depth of the objects/arrays being passed. 
This is why we are only passing in the isolated models of the component/input we are evaluating vs compare an entire filing model against an entire snapshot.

**Pull request merged**